In [2]:
#import packages needed
import requests
import json
import math


# Create the list of all ICs
ICs =['CC'
,'CSR '
,'CIT' 
,'FIC' 
,'NCATS'
,'NCCIH'
,'NCI'
,'NCRR'
,'NEI'
,'NHGRI'
,'NHLBI'
,'NIA'
,'NIAAA'
,'NIAID'
,'NIAMS'
,'NIBIB'
,'NICHD'
,'NIDA'
,'NIDCD'
,'NIDCR'
,'NIDDK'
,'NIEHS'
,'NIGMS'
,'NIMH'
,'NIMHD'
,'NINDS'
,'NINR'
,'NLM'
,'OD' ]

In [49]:
# find total records number

url = "https://api.reporter.nih.gov/v2/projects/search"
headers = {'accept': 'application/json', 'Content-Type': 'application/json'}

data = {
        "criteria":
        {
            "fiscal_years": [2021],
            "agencies":ICs
            },
       }

     #executing curl command with python codes, and transfer into json file
r = requests.post(url, data=json.dumps(data), headers=headers)
r_json = r.json()

tot_records = r_json['meta']['total']
print(f"## Number of records: {tot_records}")



## Number of records: 78895


### Hitting the 15000 Limit
78895 is far greater than 15000. If we try to fetch all records in one loop. An error message: ["System doesn't support offset value greater than 14,999. Please narrow down your search criteria."] will show.

In [28]:
# Fetch all data in one loop
meta = []
results = []

url = "https://api.reporter.nih.gov/v2/projects/search"
headers = {'accept': 'application/json', 'Content-Type': 'application/json'}

for i in range(num_calls):
    print(f'Call #{i}')
    data = {
         "criteria":
         {
             "fiscal_years": [2021],
             "agencies":ICs
             },
             "include_fields": [
                "ApplId","SubprojectId","FiscalYear","Organization", "ProjectNum","OrgCountry",
                "ProjectNumSplit","ContactPiName", "AllText","FullStudySection",
                "ProjectStartDate","ProjectEndDate"
             ],
             "offset":500*i,
             "limit":500,
             "sort_field":"project_start_date",
             "sort_order":"desc"
        }

        #executing curl command with python codes, and transfer into json file
    r = requests.post(url, data=json.dumps(data), headers=headers)
    r_json = r.json()
    
    if i == 0:
        tot_records = r_json['meta']['total']
        print(f"## Number of records: {tot_records}")
        
    if len(r_json) == 1:
        print(r_json)
        break
    else:
        if r_json['results'] != []:           
            meta.append(r_json['meta'])
            results.append(r_json['results'])
            print(f"# Number of records in this call: {len(r_json['results'])}")
        else:
            break

Call #0
## Number of records: 78870
# Number of records in this call: 500
Call #1
# Number of records in this call: 500
Call #2
# Number of records in this call: 500
Call #3
# Number of records in this call: 500
Call #4
# Number of records in this call: 500
Call #5
# Number of records in this call: 500
Call #6
# Number of records in this call: 500
Call #7
# Number of records in this call: 500
Call #8
# Number of records in this call: 500
Call #9
# Number of records in this call: 500
Call #10
# Number of records in this call: 500
Call #11
# Number of records in this call: 500
Call #12
# Number of records in this call: 500
Call #13
# Number of records in this call: 500
Call #14
# Number of records in this call: 500
Call #15
# Number of records in this call: 500
Call #16
# Number of records in this call: 500
Call #17
# Number of records in this call: 500
Call #18
# Number of records in this call: 500
Call #19
# Number of records in this call: 500
Call #20
# Number of records in this call:

### Bypassing 15000 limit by querying ICs one by one


In [53]:
url = "https://api.reporter.nih.gov/v2/projects/search"
headers = {'accept': 'application/json', 'Content-Type': 'application/json'}

all_IC_meta = [] #a list to store all retrieval meta data
all_IC_result = [] #a list to store all results from all retrievals from all ICs
# There are other ways to store data other than lists.

for IC in ICs:
    print(f'#######{IC}#######')
    meta = [] #temporary storage for each IC
    results = []
    i = 0
    r_json = {}

    while True:
    
        data = {
             "criteria":
             {
                 "fiscal_years":[2021],
               "agencies":[IC],
             },
             "include_fields": [
                "ApplId","SubprojectId","FiscalYear","Organization", "ProjectNum","OrgCountry",
                "ProjectNumSplit","ContactPiName", "AllText","FullStudySection",
                "ProjectStartDate","ProjectEndDate"
             ],
             "offset":500*i,
             "limit":500,
             "sort_field":"project_start_date",
             "sort_order":"desc"
            }

        r = requests.post(url, data=json.dumps(data), headers=headers)
        r_json = r.json()    
    
        try:
            if r_json['meta']['total'] == 0:
                print('Number of records: 0')              
                break
                
            if r_json['meta']['total']!=0:     
                meta.append(r_json['meta'])
                results.append(r_json['results'])
                print(f'Call #{i}')
                print(f"# Number of records in this call: {len(r_json['results'])}")
                i+=1
        except:
            total = 0
            for i in results:
                total += len(i)
            print(f'Number of records for {IC}:{total}')
            break

    all_IC_meta.append(meta)
    all_IC_result.append(results)
    
    



#######CC#######
Number of records: 0
#######CSR #######
Number of records: 0
#######CIT#######
Call #0
# Number of records in this call: 5
Number of records for CIT:5
#######FIC#######
Call #0
# Number of records in this call: 419
Number of records for FIC:419
#######NCATS#######
Call #0
# Number of records in this call: 500
Call #1
# Number of records in this call: 164
Number of records for NCATS:664
#######NCCIH#######
Call #0
# Number of records in this call: 362
Number of records for NCCIH:362
#######NCI#######
Call #0
# Number of records in this call: 500
Call #1
# Number of records in this call: 500
Call #2
# Number of records in this call: 500
Call #3
# Number of records in this call: 500
Call #4
# Number of records in this call: 500
Call #5
# Number of records in this call: 500
Call #6
# Number of records in this call: 500
Call #7
# Number of records in this call: 500
Call #8
# Number of records in this call: 500
Call #9
# Number of records in this call: 500
Call #10
# Number 

Call #6
# Number of records in this call: 500
Call #7
# Number of records in this call: 258
Number of records for NIMH:3758
#######NIMHD#######
Call #0
# Number of records in this call: 500
Call #1
# Number of records in this call: 452
Number of records for NIMHD:952
#######NINDS#######
Call #0
# Number of records in this call: 500
Call #1
# Number of records in this call: 500
Call #2
# Number of records in this call: 500
Call #3
# Number of records in this call: 500
Call #4
# Number of records in this call: 500
Call #5
# Number of records in this call: 500
Call #6
# Number of records in this call: 500
Call #7
# Number of records in this call: 500
Call #8
# Number of records in this call: 500
Call #9
# Number of records in this call: 500
Call #10
# Number of records in this call: 47
Number of records for NINDS:5047
#######NINR#######
Call #0
# Number of records in this call: 415
Number of records for NINR:415
#######NLM#######
Call #0
# Number of records in this call: 250
Number of rec

### Verification

In [54]:
total_all_IC_result = 0
for w in all_IC_result:
    for r in w:
        if len(r) != 0:
            total_all_IC_result += len(r)
print(f"Total records retrieved: {total_all_IC_result}")

Total records retrieved: 78895


In [35]:
total_calls = 0

for i in all_IC_meta :
        total_calls += len(i)
        
total_calls

171

In [47]:
total_calls_verify = 0
for w in all_IC_result:
     if len(w) != 0:
        total_calls_verify += len(w)
        
total_calls_verify

171

In [36]:
len(all_IC_meta)

29

In [51]:
len(all_IC_result)

29